partitionmanager/dropper_test.py

```diff
@@ -0,0 +1,159 @@
+import unittest
+from datetime import datetime, timedelta, timezone
+
+from .dropper import _drop_statement, get_droppable_partitions
+from .types import (
+    DatabaseCommand,
+    Table,
+    SqlInput,
+    SqlQuery,
+    PositionPartition,
+)
+from .types_test import mkPPart, mkTailPart, mkPos
+
+
+def _timestamp_rsp(year, mo, day):
+    return [
+        {"UNIX_TIMESTAMP": datetime(year, mo, day, tzinfo=timezone.utc).timestamp()}
+    ]
+
+
+class MockDatabase(DatabaseCommand):
+    def __init__(self):
+        self._responses = list()
+        self.num_queries = 0
+
+    def add_response(self, expected, response):
+        self._responses.insert(0, {"expected": expected, "response": response})
+
+    def run(self, cmd):
+        self.num_queries += 1
+        if not self._responses:
+            raise Exception(f"No mock responses available for cmd [{cmd}]")
+
+        r = self._responses.pop()
+        if r["expected"] in cmd:
+            return r["response"]
+
+        raise Exception(f"Received command [{cmd}] and expected [{r['expected']}]")
+
+    def db_name(self):
+        return SqlInput("the-database")
+
+
+class TestDropper(unittest.TestCase):
+    def test_drop_statement_empty(self):
+        table = Table("burgers")
+        parts = []
+        with self.assertRaises(ValueError):
+            _drop_statement(table, parts)
+
+    def test_drop_statement(self):
+        table = Table("burgers")
+        parts = [PositionPartition("p_start")]
+        self.assertEqual(
+            _drop_statement(table, parts),
+            "ALTER TABLE `burgers` DROP PARTITION IF EXISTS `p_start`;",
+        )
+
+    def test_get_droppable_partitions_invalid_config(self):
+        database = MockDatabase()
+        table = Table("burgers")
+        partitions = [PositionPartition("p_start")]
+        current_timestamp = datetime(2021, 1, 1, tzinfo=timezone.utc)
+        current_position = PositionPartition("p_20210102").set_position([10])
+
+        with self.assertRaises(ValueError):
+            get_droppable_partitions(
+                database, partitions, current_position, current_timestamp, table
+            )
+
+    def test_get_droppable_partitions(self):
+        database = MockDatabase()
+        database.add_response("WHERE `id` > '100'", _timestamp_rsp(2021, 5, 20))
```

Comment:
I think this is OK, actually. The responses here have to be ordered in the same way as the tests below because of how the database mock works, not because of some underlying logic requirement. For each of the actual `get_droppable_partitions` calls, the query is responsible for returning only one row (and one column) so there's no possibility for misordering. 

That said the `partitions` argument to `get_droppable_partitions` does need to hold a certain order, which I'm adding in f4d546e.